synapse/handlers/room.py

```diff
@@ -91,66 +100,94 @@ def upgrade_room(self, requester, old_room_id, new_version):
 
         user_id = requester.user.to_string()
 
-        with (yield self._upgrade_linearizer.queue(old_room_id)):
-            # start by allocating a new room id
-            r = yield self.store.get_room(old_room_id)
-            if r is None:
-                raise NotFoundError("Unknown room id %s" % (old_room_id,))
-            new_room_id = yield self._generate_room_id(
-                creator_id=user_id, is_public=r["is_public"],
-            )
+        # Check if this room is already being upgraded by another person
+        for key in self._upgrade_response_cache.pending_result_cache:
+            if key[0] == old_room_id and key[1] != user_id:
```

Comment:
We probably also want to block the same user if  they try to upgrade to different room versions. Currently the response cache below lets the same user upgrade to different versions at the same time.